src/inmanta/data/__init__.py

```diff
@@ -402,41 +405,395 @@ def as_end_filter(
+        if len(filter_elements) == 1:
+            col_name, coll_type, value = filter_elements[0]
+            filter = f"{coll_type.get_accessor(col_name)} {relation} {ac(value)}"
+            return [filter], ac.args
+        else:
+            # composed filter:
+            # 1. comparison of two tuples (c_a, c_b) < (c_a, c_b)
+            # 2. nulls must be removed to get proper comparison
+            names_tuple = ", ".join([coll_type.get_accessor(col_name) for col_name, coll_type, value in filter_elements])
+            values_references_tuple = ", ".join(
+                [coll_type.coalesce_to_min(ac(value)) for col_name, coll_type, value in filter_elements]
+            )
+            filter = f"({names_tuple}) {relation} ({values_references_tuple})"
+            return [filter], ac.args
+
+    def get_order_elements(self, invert: bool) -> List[Tuple[ColumnNameStr, ColumnType, PagingOrder]]:
+        """
+        return a list of column/column type/order triples, to format an ORDER BY or FILTER statement
+        """
+        order = self.get_order(invert)
+        id_name, id_type = self.id_column
+        return [
+            (self.order_by_column, self.get_valid_sort_columns()[self.order_by_column], order),
+            (id_name, id_type, order),
+        ]
+
+    def get_paging_boundaries(self, first: dict[str, object], last: dict[str, object]) -> PagingBoundaries:
+        """Return the page boundaries, given the first and last record returned"""
+        if self.get_order() == PagingOrder.ASC:
+            first, last = last, first
+
+        order_column_name = self.order_by_column
+        order_type: ColumnType = self.get_valid_sort_columns()[self.order_by_column]
+
+        id_column, id_type = self.id_column
+
+        def assert_not_null(in_value: Optional[PRIMITIVE_SQL_TYPES]) -> PRIMITIVE_SQL_TYPES:
+            # Make mypy happy
+            assert in_value is not None
+            return in_value
+
+        return PagingBoundaries(
+            start=assert_not_null(order_type.get_value(first[order_column_name])),
+            first_id=assert_not_null(id_type.get_value(first[id_column])),
+            end=assert_not_null(order_type.get_value(last[order_column_name])),
+            last_id=assert_not_null(id_type.get_value(last[id_column])),
+        )
+
+
+class VersionedResourceOrder(AbstractDatabaseOrderV2):
+    """Represents the ordering by which resources should be sorted"""
+
+    @classmethod
+    def get_valid_sort_columns(cls) -> Dict[ColumnNameStr, ColumnType]:
+        return {
+            ColumnNameStr("resource_type"): StringColumn,
+            ColumnNameStr("agent"): StringColumn,
+            ColumnNameStr("resource_id_value"): StringColumn,
+        }
+
+    @property
+    def id_column(self) -> Tuple[ColumnNameStr, ColumnType]:
         """Name of the id column of this database order"""
-        return ColumnNameStr("resource_version_id")
+        return ColumnNameStr("resource_id"), StringColumn
```

Comment:
Why isn't this in `get_valid_sort_columns()` as well? Same question for all concrete order classes.

Reply:
it is not in itself a valid sort column, i.e. the user can not provide it.